File: leetcode/1104.cpp
#include <vector>
#include <iostream>
#include <cmath>
#include <algorithm>

using namespace std;

class Solution {
public:
  vector<int> pathInZigZagTree(int label) {
    vector<int> ans;
    // labal所处行数
    int level = log2(label);
    while (label > 1) {
      ans.push_back(label);
      label = 3 * (1 << (--level)) - (label >> 1) - 1;
    }
    ans.push_back(1);
    reverse(ans.begin(), ans.end());
    return ans;
  }
};

int main() {
  vector<int> ans = Solution().pathInZigZagTree(26);
  for_each(ans.begin(), ans.end(), [](int val) { cout << val << " "; });
  cout << endl;
  return 0;
}

File: leetcode/1104.py
from typing import List


class Solution:

    def pathInZigZagTree(self, label: int) -> List[int]:
        tree, num, level = [], 0, 0

        while num < label:
            nums = pow(2, level)
            if level % 2 == 0:
                for i in range(num + 1, num + nums + 1, 1):
                    tree.append(i)
                    if i == label:
                        break
            else:
                for i in range(num + nums, num, -1):
                    tree.append(i)
                    if i == label:
                        break
            num += nums
            level += 1

        ans, index = [], len(tree) - 1
        while index > 0:
            ans.append(tree[index])
            index = (index - 1) >> 1
        ans.append(1)
        ans.reverse()
        return ans
